python/ray/serve/backend_state.py

```diff
@@ -296,67 +288,64 @@ def __init__(self, controller_name: str, detached: bool,
         self._replica_tag = replica_tag
         self._backend_tag = backend_tag
         self._version = version
         self._start_time = None
         self._prev_slow_startup_warning_time = None
 
-        self._state = ReplicaState.SHOULD_START
-
     def __get_state__(self) -> Dict[Any, Any]:
         return self.__dict__.copy()
 
     def __set_state__(self, d: Dict[Any, Any]) -> None:
         self.__dict__ = d
-        self._recover_from_checkpoint()
-
-    def _recover_from_checkpoint(self) -> None:
-        if self._state == ReplicaState.STARTING_OR_UPDATING:
-            # We do not need to pass in the class here because the actor
-            # creation has already been started if this class was checkpointed
-            # in the STARTING_OR_UPDATING state.
-            self.start_or_update()
-        elif self._state == ReplicaState.STOPPING:
-            self.stop()
 
     @property
     def replica_tag(self) -> ReplicaTag:
         return self._replica_tag
 
+    @property
+    def backend_tag(self) -> BackendTag:
+        return self._backend_tag
+
     @property
     def version(self):
         return self._version
 
     @property
     def actor_handle(self) -> ActorHandle:
-        assert self._state is not ReplicaState.SHOULD_START, (
-            f"State must not be {ReplicaState.SHOULD_START}")
         return self._actor.actor_handle
 
-    def start_or_update(self, backend_info: BackendInfo,
-                        version: BackendVersion) -> None:
-        """Transition from SHOULD_START -> STARTING_OR_UPDATING.
-
-        Should handle the case where it's already STARTING_OR_UPDATING.
+    def start(self, backend_info: BackendInfo, version: BackendVersion):
         """
-        assert self._state in {
-            ReplicaState.SHOULD_START,
-            ReplicaState.STARTING_OR_UPDATING,
-            ReplicaState.RUNNING,
-        }, (f"State must be {ReplicaState.SHOULD_START}, "
-            f"{ReplicaState.STARTING_OR_UPDATING}, or"
-            f"{ReplicaState.RUNNING}, *not* {self._state}.")
-
-        self._actor.start_or_update(backend_info)
+        Start a new actor for current BackendReplica instance.
+        """
+        self._actor.start(backend_info, version)
         self._start_time = time.time()
         self._prev_slow_startup_warning_time = time.time()
-        self._state = ReplicaState.STARTING_OR_UPDATING
         self._version = version
 
+    def update(self, user_config: Any):
+        """
+        Update user config of existing actor behind current
+        BackendReplica instance.
+        """
+        self._actor.update(user_config)
+        self._start_time = time.time()
+        self._version = BackendVersion(
+            self._version.code_version, user_config=user_config)
+
+    def recover(self):
+        """
+        Recover states in BackendReplica instance by fetching running actor
+        status
+        """
+        self._actor.recover()
+        self._start_time = time.time()
```

Comment:
Can we explicitly set `self._version` to whatever it should be here? And leave a comment saying we'll fetch it from the replica dynamically.

Reply:
it will be `self._version = self._version` though since it has to come from start() or update_user_config() and both will set or re-compute a version and assign to `self._version`